test/robot.go

```diff
@@ -22,20 +22,25 @@ func init() {
 
 func robot(ctx context.Context, t model.KeyedTest, sub model.KeyedSubmission, ball io.Reader) (err error) {
 	log.Debugf(ctx, "Executing robot tester")
-	var testMap, stdin io.ReadCloser
+	var testMap io.ReadCloser
 	if testMap, err = util.Load(util.CloudContext(ctx), util.TemplateBucket, t.Params["tests"]); err != nil {
 		return
 	}
 	defer testMap.Close()
 
-	if stdin, err = util.Load(util.CloudContext(ctx), sub.Code.Bucket, path.Dir(sub.Code.Name)+"/"+util.FileNames["robot"]); err != nil {
-		return
-	}
-	defer stdin.Close()
-
 	var testMapBytes, stdinBytes []byte
-	if stdinBytes, err = ioutil.ReadAll(stdin); err != nil {
-		return
+	tr := tar.NewReader(ball)
+	for {
```

Comment:
That loop looks a bit crooked. You expect only one file, so you could do `tr.Next()` expect no `io.EOF`, then process stuff, then do `tr.Next()` again and expect `io.EOF`. Anyway, if there are multiple files and no error you'll write to `stdinBytes` multiple times which could be confusing to debug.